src/univers/version_range.py

```diff
@@ -2,73 +2,397 @@
-from univers.versions import version_class_by_scheme
-from univers.versions import validate_scheme
+from univers.version_constraint import VersionConstraint
 
 
+@attr.s(frozen=True, order=False, eq=True, hash=True)
 class VersionRange:
-    # one of <> >= =< or != or =
-    operator = ""
-    version = None
+    """
+    Base version range class. Subclasses must provide implememt.
+    """
 
-    def __init__(self, version_range_string, scheme):
-        version_range_string = remove_spaces(version_range_string)
-        self.operator, self.version = self.split(version_range_string)
+    # Versioning scheme. By convention this is the same as the Package URL
+    # package type since this "defines" most commonly the versioning scheme of a
+    # package, such as "npm" (whose scheme is defined in the "node-semver" npm
+    # package. This could be something else though and a purl may be accompanied
+    # by a version range for another scheme; for example, a
+    # ``pkg:github/foo/bar`` purl could be accompanied by a a ``vers:npm/12.3``
+    # range. Subclasses MUST provide this.
+    scheme = None
 
-        try:
-            validate_scheme(scheme)
-            self.validate()
-        except:
-            raise ValueError(f"Version range{version_range_string} has no bounds")
+    # Version subclass to use with this versioning scheme, such as
+    # PypiVersion. Subclasses MUST provide this.
+    version_class = None
 
-        version_class = version_class_by_scheme[scheme]
+    # A list of lists of VersionConstraint where the outer list is an "OR" of
+    # the innner lists that are each "ANDs" of atomic constraints
+    constraints = attr.ib(type=list, default=attr.Factory(list))
 
-        self.version = version_class(self.version)
+    def __attrs_post_init__(self, *args, **kwargs):
+        VersionConstraint.sort(self.constraints)
 
-    def validate(self):
-        # self.operator will always have a valid value
-        if not self.version:
-            raise ValueError()
+    @classmethod
+    def from_native_range(cls, string):
+        """
+        Return a VersionRange built from a scheme-specific, native version range
+        ``string``. Subclasses must implement.
+        """
+        return NotImplementedError
 
-    @staticmethod
-    def split(version_range):
+    def to_native_range(self):
         """
-        Return a tuple of (operator, range value) given a version ``range``
-        string such as ">=2.3".
+        Return a native range string for this VersionRange. Subclasses must
+        implement.
         """
-        operators = ">=", "<=", "!=", "<", ">", "="
-        for operator in operators:
-            if version_range.startswith(operator):
-                return operator, version_range.lstrip("><=!")
+        return NotImplementedError
 
-        # Contains no operator, so assume equality
-        return "=", version_range
+    @classmethod
+    def from_vers(cls, vers):
```

Comment:
Why is this not named "from_string" ? This is kinda similar to the packageurl's from_string method

Reply:
fixed in latest commit: renamed to "from_string"